### PR TITLE
initial Apache SkyWalking tracer support

### PR DESCRIPTION
**Pull request by Shikugawa**
This PR introduces initial Apache SkyWalking support. This PR depends on istio-proxy which includes these fixes (https://github.com/envoyproxy/envoy/pull/16203, https://github.com/envoyproxy/envoy/pull/16264).

Screen Shot 2021-05-02 at 21 02 46

Screen Shot 2021-05-02 at 21 03 10

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[x] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Review by shamsher31 — COMMENTED** [review]
@Shikugawa Can you please add release notes for this PR?

File: pilot/docker/Dockerfile.proxyv2 [hunk 1]
```diff
@@ -30,7 +30,7 @@ COPY gcp_envoy_bootstrap.json /var/lib/istio/envoy/gcp_envoy_bootstrap_tmpl.json
 # Install Envoy.
 COPY $SIDECAR /usr/local/bin/$SIDECAR
 
-# Environment variable indicating the exact proxy sha - for debugging or version-specific configs 
+# Environment variable indicating the exact proxy sha - for debugging or version-specific configs
```
Comment (howardjohn) [inline pilot/docker/Dockerfile.proxyv2:33]: nit: remove this change, its not related. But its a good change in general

File: istioctl/cmd/dashboard.go [hunk 2]
```diff
@@ -365,6 +365,42 @@ func controlZDashCmd() *cobra.Command {
 	return cmd
 }
 
+// port-forward to SkyWalking UI on istio-system
+func skywalkingDashCmd() *cobra.Command {
+	var opts clioptions.ControlPlaneOptions
+	cmd := &cobra.Command{
+		Use:   "skywalking",
+		Short: "Open SKyWalking UI",
```
Comment (douglas-reid) [inline istioctl/cmd/dashboard.go:373]:
```suggestion
		Short: "Open SkyWalking UI",
```

File: istioctl/cmd/dashboard.go [hunk 3]
```diff
@@ -365,6 +365,42 @@ func controlZDashCmd() *cobra.Command {
...
+		Long:  "Open SkyWalking UI on Istio",
```
Comment (douglas-reid) [inline istioctl/cmd/dashboard.go:374]:
```suggestion
		Long:  "Open the Istio dashboard in the SkyWalking UI",
```
?

File: pilot/pkg/networking/core/v1alpha3/tracing_test.go [hunk 4]
```diff
@@ -101,6 +105,40 @@ func TestConfigureTracing(t *testing.T) {
...
+		clusterLookupFn = extensionproviders.LookupCluster
+	}()
+
+	testcases := []struct {
+		name     string
+		spec     *tpb.Telemetry
+		opts     buildListenerOpts
+		expected *hpb.HttpConnectionManager_Tracing
+	}{
+		{
```
Comment (douglas-reid) [inline pilot/pkg/networking/core/v1alpha3/tracing_test.go:125]: Is there any reason this couldn't be added as a testcase to the existing test?  It looks like the setup is the same, as is the actually test loop.

Comment (Shikugawa) [reply pilot/pkg/networking/core/v1alpha3/tracing_test.go:125]: As it is to be more clear to separate Zipkin and SkyWalking test.

File: releasenotes/notes/32588.yaml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+apiVersion: release-notes/v2
+kind: feature
+area: telemetry
+
+issue:
+- https://github.com/istio/istio/pull/32588
+
+releaseNotes:
+- |
+  **Added** Introduced to deliver Apache SkyWalking tracer configuration to envoy.
```
Comment (shamsher31) [inline releasenotes/notes/32588.yaml:10]:
```suggestion
  **Added** support for [Apache SkyWalking](https://skywalking.apache.org/) tracer. Now you can run the `istioctl dashboard skywalking` command to view skywalking dashboard UI.
```

File: pilot/pkg/networking/core/v1alpha3/tracing_test.go (hunk 4 above)
Comment (wu-sheng) [reply pilot/pkg/networking/core/v1alpha3/tracing_test.go:125]: It is better to keep these two separated. In case of new features add in the future.

Comment (douglas-reid) [reply pilot/pkg/networking/core/v1alpha3/tracing_test.go:125]: But the other test isn't meant to be *just* a zipkin test. It is a test of generating tracing configuration. The fact that zipkin is used was just to have a placeholder fake provider. I'd much prefer to add SkyWalking to the list of fake providers and keep a single test for validating configuration than have separate repetitive tests for each provider.

If the time comes when that is no longer tenable, we can split things out then.

Comment (Shikugawa) [reply pilot/pkg/networking/core/v1alpha3/tracing_test.go:125]: Sure. I will add other providers for extension provider tests in the future (not in this PR). I will fix it just to add existing placeholder and add/separate with appending other tests such as datadog, opencensus agent... as an improvement of new Telemetry API.

**Review by douglas-reid — APPROVED** [review]

**Comment by Shikugawa** [comment]
@istio/repo-admins Could anyone merge this?

**Comment by Shikugawa** [comment]
@howardjohn @shamsher31 Could you approve this?

**Review by shamsher31 — APPROVED** [review]
Thanks.

**Review by howardjohn — APPROVED** [review]

**Comment by du2016** [comment]
@Shikugawa i have build myself and use bookinfo to test,but the trace in skywalking only have one span,the img you upload have show the same problem: all of the service is called by user,i have dump the envoy config file , not got any problem,please have a look ,thank you

**Comment by wu-sheng** [comment]
@du2016 I think bookinfo doesn't propagate sw8 header? Or have SkyWalking agents installed?

**Comment by Shikugawa** [comment]
> @Shikugawa i have build myself and use bookinfo to test,but the trace in skywalking only have one span,the img you upload have show the same problem: all of the service is called by user,i have dump the envoy config file , not got any problem,please have a look ,thank you

@du2016 Could you show me the detailed reproduction step?

**Comment by du2016** [comment]
@wu-sheng envoy is not support sw8-x header ,https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/trace/v3/skywalking.proto.html?highlight=sw8
@Shikugawa i have install skywalking in my k8s cluster,and config extensionProviders in meshconfig as blow
```
    extensionProviders:
    - skywalking:
        service: skywalking-oap.skywalking.svc.cluster.local
        port: 11800
      name: envoy.tracers.skywalking
```

then  add a Telemetry cr 
```
apiVersion: telemetry.istio.io/v1alpha1
kind: Telemetry
metadata:
  name: nginx
  namespace: default
spec:
  tracing:
  - customTags:
      my_new_foo_tag:
        literal:
          value: foo
    providers:
    - name: envoy.tracers.skywalking
    randomSamplingPercentage: 100
```

config istio gateway and use curl to got traffic,view the skywalking ui,

topology:

trace:


I just checked the envoy doc, this problem mybe caused by istio is not config `start_child_span` in route filter,start_child_span must set to true when use SkyWalking tracer  in envoy doc,

then i add start_child_span to router filter,but also not got full trace

**Comment by wu-sheng** [comment]
> envoy is not support sw8-x header ,https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/trace/v3/skywalking.proto.html?highlight=sw8

We only need `sw8` header to propagate. not `sw8-x` for this stage. And this support is for the demo app(bookinfo), rather than the envoy. Envoy tracer has taken care of it already.

**Comment by du2016** [comment]
@wu-sheng ,thanks ，i'll have a try

**Comment by du2016** [comment]
@Shikugawa  @wu-sheng ,i have propagate sw8 header in my demo app,and also should set start_child_span to true in istio

**Comment by wu-sheng** [comment]
> @Shikugawa  @wu-sheng ,i have propagate sw8 header in my demo app,and also should set start_child_span to true in istio, then work's ok

Could you update Istio doc somewhere accordingly?

**Comment by du2016** [comment]
yes,i'd like to work on this

**Comment by Shikugawa** [comment]
@douglas-reid @howardjohn I think that we should setup `start_child_span` manually is pretty weird. So we should setup it automatically. But It may need a pretty refactor of building `HTTPConnManager`. My plan is to introduce `HttpConnManagerContext` which propagates some information (in this situation, whether to setup `start_child_span` or not) across `HTTPConnManager` building phase. WDYT?

**Comment by Shikugawa** [comment]
@du2016 @wu-sheng Hi. I fixed this problem with internal code change. With this, we don't have to take care about `start_child_span` with SkyWalking provider.

**Comment by Elven-coder** [comment]
@wu-sheng Hi Wu Sheng. Could Istio 1.8.2 and sky 8.1.0 support tracer?  The tracer page and profile page have no data in my skywalking dashboard. 

skywalking-0
skywalking-1
skywalking-2
skywalking-3

**Comment by Shikugawa** [comment]
> @wu-sheng Hi Wu Sheng. Could Istio 1.8.2 and sky 8.1.0 support tracer? The tracer page and profile page have no data in my skywalking dashboard.
> 
> skywalking-0
> 
> skywalking-1
> 
> skywalking-2
> 
> skywalking-3

SkyWalking 8.1.0 is no supported.

**Comment by wu-sheng** [comment]
@Elven-coder also, the profile requires a language agent installed in your service, it works in in-process level, which can't be read from sidecar(envoy).

**Comment by Elven-coder** [comment]
Got it. Thanks.

**Comment by funlake** [comment]
> du2016

Hi,what was the skywalking version and istio verison you tested?just try on my k8s and get no luck.

**Comment by Shikugawa** [comment]
> Hi,what was the skywalking version and istio verison you tested?just try on my k8s and get no luck.

It only supports 8.4.0 >= because of API compatibility.

**Comment by funlake** [comment]
> It only supports skywalking >= 8.4.0 because of API compatibility. And you should use istio >= 1.11.

Thank you ,so do we have any docs about how to make this done ? i've tried istio1.12 & sw 8.8.1, metrics & access log  indeed showed expected data in skywalking ,but tracing page still empty.

my meshConfig 
-----------------
```
apiVersion: v1
data:
  mesh: |-
    extensionProvider:
    - skywalking:
        service: skywalking-oap.skywalking.svc.cluster.local
        port: 11800
      name: envoy.tracers.skywalking
    accessLogFile: /dev/stdout
    defaultConfig:
      discoveryAddress: istiod.istio-system.svc:15012
      envoyAccessLogService:
        address: skywalking-oap.skywalking.svc:11800
      envoyMetricsService:
        address: skywalking-oap.skywalking.svc:11800
...
```

here is my telmetry config
------------
```
apiVersion: telemetry.istio.io/v1alpha1
kind: Telemetry
metadata:
  name: tracesky
  namespace: bookinfo
spec:
  tracing:
  - providers:
    - name: "envoy.tracers.skywalking"
    randomSamplingPercentage: 100
```

**Comment by Shikugawa** [comment]
Does it work correctly by setting `defaultProviders`?

```
apiVersion: v1
data:
  meshNetworks: |-
    networks: {}
  mesh: |-
    ...
    defaultProviders:
      tracing: "skywalking"
```

**Comment by funlake** [comment]
Thanks but still no luck
my full istio config below,and already restarted istiod & bookinfo apps after updated it
---------------------
```
apiVersion: v1
data:
  mesh: |-
    extensionProvider:
    - skywalking:
        service: skywalking-oap.skywalking.svc.cluster.local
        port: 11800
      name: envoy.tracers.skywalking
    accessLogFile: /dev/stdout
    defaultConfig:
      defaultProviders:
        tracing: "skywalking"
      discoveryAddress: istiod.istio-system.svc:15012
      envoyAccessLogService:
        address: skywalking-oap.skywalking.svc:11800
      envoyMetricsService:
        address: skywalking-oap.skywalking.svc:11800
      proxyMetadata: {}
      proxyStatsMatcher:
        inclusionRegexps:
        - .*membership_healthy.*
        - .*upstream_cx_active.*
        - .*upstream_cx_total.*
        - .*upstream_rq_active.*
        - .*upstream_rq_total.*
        - .*upstream_rq_pending_active.*
        - .*lb_healthy_panic.*
        - .*upstream_cx_none_healthy.*
      tracing:
        zipkin:
          address: jaeger01-collector.jaeger:9411
    enableEnvoyAccessLogService: true
    enableTracing: true
    enablePrometheusMerge: true
    rootNamespace: istio-system
    trustDomain: cluster.local
  meshNetworks: 'networks: {}'
kind: ConfigMap
metadata:
  labels:
    install.operator.istio.io/owning-resource: unknown
    install.operator.istio.io/owning-resource-namespace: istio-system
    istio.io/rev: default
    operator.istio.io/component: Pilot
    operator.istio.io/managed: Reconcile
    operator.istio.io/version: 1.12.1
    release: istio
  name: istio
  namespace: istio-system
```

**Comment by douglas-reid** [comment]
There is a naming mismatch. If you want `skywalking` to be the default provider, you'll need to use the same name consistently in the config.

Either:

```
extensionProvider:
- skywalking:
     service: skywalking-oap.skywalking.svc.cluster.local
     port: 11800
  name: envoy.tracers.skywalking
defaultConfig:
  defaultProviders:
    tracing: envoy.tracers.skywalking
```

or:

```
extensionProvider:
- skywalking:
     service: skywalking-oap.skywalking.svc.cluster.local
     port: 11800
  name: skywalking
defaultConfig:
  defaultProviders:
    tracing: skywalking
```

You may also want to remove the:

```
tracing:
  zipkin:
     address: jaeger01-collector.jaeger:9411
```

**Comment by funlake** [comment]
> There is a naming mismatch. If you want `skywalking` to be the default provider, you'll need to use the same name consistently in the config.
> 
> Either:
> 
> ```
> extensionProvider:
> - skywalking:
>      service: skywalking-oap.skywalking.svc.cluster.local
>      port: 11800
>   name: envoy.tracers.skywalking
> defaultConfig:
>   defaultProviders:
>     tracing: envoy.tracers.skywalking
> ```
> 
> or:
> 
> ```
> extensionProvider:
> - skywalking:
>      service: skywalking-oap.skywalking.svc.cluster.local
>      port: 11800
>   name: skywalking
> defaultConfig:
>   defaultProviders:
>     tracing: skywalking
> ```
> 
> You may also want to remove the:
> 
> ```
> tracing:
>   zipkin:
>      address: jaeger01-collector.jaeger:9411
> ```

Thanks.
Istio config  now （access log indeed works,but tracing still show nothings）：

=====================
```
apiVersion: v1
data:
  mesh: |-
    extensionProvider:
    - skywalking:
        service: skywalking-oap.skywalking.svc.cluster.local
        port: 11800
      name: envoy.tracers.skywalking
    accessLogFile: /dev/stdout
    defaultConfig:
      defaultProviders:
        tracing: envoy.tracers.skywalking
      discoveryAddress: istiod.istio-system.svc:15012
      envoyAccessLogService:
        address: skywalking-oap.skywalking.svc:11800
        #address: als-server.amt:8080
      #envoyMetricsService:
      #  address: skywalking-oap.skywalking.svc:11800
      proxyMetadata: {}
      proxyStatsMatcher:
        inclusionRegexps:
        - .*membership_healthy.*
        - .*upstream_cx_active.*
        - .*upstream_cx_total.*
        - .*upstream_rq_active.*
        - .*upstream_rq_total.*
        - .*upstream_rq_pending_active.*
        - .*lb_healthy_panic.*
        - .*upstream_cx_none_healthy.*
      #tracing:
      #  zipkin:
      #    #address: zipkin.istio-system:9411
      #    address: jaeger01-collector.jaeger:9411
    enableEnvoyAccessLogService: true
    enableTracing: true
    enablePrometheusMerge: true
    rootNamespace: istio-system
    trustDomain: cluster.local
  meshNetworks: 'networks: {}'
kind: ConfigMap
metadata:
  labels:
    install.operator.istio.io/owning-resource: unknown
    install.operator.istio.io/owning-resource-namespace: istio-system
    istio.io/rev: default
    operator.istio.io/component: Pilot
    operator.istio.io/managed: Reconcile
    operator.istio.io/version: 1.12.1
    release: istio
  name: istio
  namespace: istio-system
```

**Comment by funlake** [comment]
Tried with istio 1.12.5, finally get tracing records,but results just showed  wierd.so did i miss any configuration?
skywalking

**Comment by funlake** [comment]
> Tried with istio 1.12.5, finally get tracing records,but results just showed wierd.i think the traces just broken, none of single trace showing full chain .so did i miss any configuration? skywalking

Propagate sw8 header in my demo codes and everything works ok.

**Comment by stevenlii** [comment]
> Istio doc

Hi @du2016, did  you update Istio doc somewhere?

**Comment by kezhenxu94** [comment]
> Hi @du2016, did you update Istio doc somewhere?
> 
> > yes,i'd like to work on this

Since @du2016 hasn't updated this for a long time, I plan to update the related doc, [here](https://github.com/istio/istio/pull/40269) is the first step to add a sample configuration. (next I will add the doc in istio.io)

**Pull request merged**